### DISCUSSION TITLE
Alt/Option on Mac

### DISCUSSION THREAD

**Discussion by eric-hemasystems** [orbitalquark/textadept#273] [Q&A] [answered]
I'm trying to make some shortcuts that basically run `Ctrl+Alt+1`....`Ctrl+Alt+9` for each number on the keyboard. This works great on Linux but on Mac I can't seem to activate the shortcut. On Mac the `Alt` key is labeled `Option` so I should press should be `Ctrl`, `Option`, and some number.

I've confirmed that my shortcut is installed by hitting `Ctrl-E` to bring up the Lua input and then typing `keys['Ctrl+Alt+3']()` and it executes just fine. Is there something preventing the Option/Alt key from working? I tried some other predefined shortcuts like "Next View" (`Option`, `Shift`, `Right`) and that also seemed to not work.

**Comment by orbitalquark** [comment]
Whenever I have issues with keybindings, I temporarily edit *core/keys.lua* to show keybindings in the statusbar as I type them: https://github.com/orbitalquark/textadept/blob/1dbcff60ec070a1881090ce8669ff1214244188f/core/keys.lua#L180

  ui.statusbar_text = string.format('"%s"', key_seq)

The Alt/option key on macOS is represented by 'alt': https://orbitalquark.github.io/textadept/api.html#key-sequences. Also, key modifiers are case-sensitive. Make sure you are using 'ctrl+alt' for example.

**Reply by eric-hemasystems** [reply]
> so, key modifiers are case-sensitive

Sorry that was a typo in my message. My code actually was using all lowercase.

> `ui.statusbar_text = string.format('"%s"', key_seq)`

That is brilliant for debugging! Unfortunately I'm still lost.

If I type key combos that use either cmd or ctrl then they show up in the status bar:

![Screen Shot 2022-10-05 at 10 12 12](https://user-images.githubusercontent.com/34583914/194082048-2a1af9c2-a365-461f-a67e-70e918d38a90.png)

But if I try the same thing with the option button I get some unicode character that looks like the status bar cannot represent:

![Screen Shot 2022-10-05 at 10 13 16](https://user-images.githubusercontent.com/34583914/194082270-863f2c72-c255-4a33-91d5-abad657156a0.png)

Furthermore in the actual editor it looks like the unicode character shows up and it is a `¡` which is the [inverted exclamation mark](https://en.wikipedia.org/wiki/Inverted_question_and_exclamation_marks). In fact if you look at that wiki page it says on Mac you type that character with Option+1. Looking at [the wiki page on the mac option key](https://en.wikipedia.org/wiki/Option_key) it seems pretty much all keys have some sort of unicode character they type with the `Option` key and it seems that is taking priority over any key combo I program into Textadept.

I tried to be clever and set the key combo to `ctrl-¡` or even just `¡` but that didn't work either. It seems really that the option key on a Mac is not really like the Alt key since it already has predefined behavior that is taking priority. Maybe there is something within Gtk that would allow Textadept to establish priority and only fallback on the OS behavior if there is no key combo defined in Textadept for the option?

**Comment by orbitalquark** [comment]
Hmm. Thanks for the info. I will have to investigate this when I have some time.

**Comment by orbitalquark** [comment, marked as answer]
After investigating this, Apple's Alt/Option key functions as a compose key and would insert an extended ascii character (which is invalid UTF-8 and displays as a weird character in the statusbar). I ended up uncommenting the "print(code, M.KEYSYMS[code], shift, control, alt, cmd, caps)" statement in *core/keys.lua* and ran Textadept from the terminal in order to find the codes being emitted by macOS. For example, if 141 is emitted, then you'd use `keys[string.char(141)] = ...` to bind that key to an action. Note that Textadept removes the 'alt' modifier for composed keys.

**Reply by eric1234** [reply]
K, so that gives me a work-around but what about default keybindings that use `alt` on Mac? Examples such as `⌥⇢` to word right or `⌥⇠` for word left?

**Comment by orbitalquark** [comment]
Those should be fine. Textadept only removes the 'alt' modifier for composed keys.
